### PR TITLE
VIDEO-5226: send auto track switch off hints bit delayed. 

### PR DESCRIPTION
**Pull request by makarandp0**
with client track switch off control set to auto - sdk request that track be turned off when it determines that track is not visible. However we found that sometimes react apps ends up causing elements to get hidden momentarily when layout changes. This results in  turn off and turn on request for the track in quick succession, causing a black screen momentarily. 

This fix delays requesting track turn off request by 50 ms. This required some refactoring as previously track switchOn/SwitchOff requests were clubbed together with render dimension changes. now they are handled separately.

### CODE REVIEW

**Review by PikaJoyce — APPROVED** [review]
Woo! LGTM!

File: lib/media/track/remotevideotrack.js [hunk 1]
```diff
@@ -4,8 +4,10 @@ const mixinRemoteMediaTrack = require('./remotemediatrack');
 const VideoTrack = require('./videotrack');
 const documentVisibilityMonitor = require('../../util/documentvisibilitymonitor.js');
 const { NullObserver } = require('../../util/nullobserver.js');
+const Timeout = require('../../util/timeout');
 
 const RemoteMediaVideoTrack = mixinRemoteMediaTrack(VideoTrack);
+const TRACK_TURN_OF_DELAY_MS = 50;
```
Comment (charliesantos) [inline lib/media/track/remotevideotrack.js:10]: Should we put this in constants.js file? Also, if my understanding is correct, this implementation is essentially debouncing right? I think 50ms is not enough. I can see the turn on/off behavior if I scroll the participants list on/off screen too fast. How about 500ms?

**Review by charliesantos — CHANGES_REQUESTED** [review]
Nice solution Mak! I left a few questions.

File: lib/media/track/remotevideotrack.js [hunk 2]
```diff
@@ -291,29 +292,37 @@ function setupDocumentVisibilityTurnOff(removeVideoTrack) {
...
-  let elements = removeVideoTrack._getAllAttachedElements();
-  let updatedRenderHint = { };
-  if (removeVideoTrack._clientTrackSwitchOffControl === 'auto') {
-    // consider only visible elements.
-    elements = elements.filter(el => !removeVideoTrack._invisibleElements.has(el));
-    updatedRenderHint.enabled = document.visibilityState === 'visible' && elements.length > 0;
+  const visibleElements = removeVideoTrack._getAllAttachedElements().filter(el => !removeVideoTrack._invisibleElements.has(el));
+  const enabled = document.visibilityState === 'visible' && visibleElements.length > 0;
+
+  if (enabled === true) {
```
[the remark this answers is not in the document]
Comment (makarandp0) [reply lib/media/track/remotevideotrack.js:303]: I think explicite check for boolean is generally better. But we do not follow this rule everywhere (IMO we should)

[the remark this answers is not in the document]
Comment (charliesantos) [reply lib/media/track/remotevideotrack.js:303]: I don't have a strong preference. We can leave it as is.

File: lib/media/track/remotevideotrack.js [hunk 3]
```diff
@@ -291,29 +292,37 @@ function setupDocumentVisibilityTurnOff(removeVideoTrack) {
...
+    removeVideoTrack._turnOffTimer.clear();
+    removeVideoTrack._setRenderHint({ enabled: true });
```
[the remark this answers is not in the document]
Comment (makarandp0) [reply lib/media/track/remotevideotrack.js:305]: Although RemoteVideoTrack may send multiple enable=true requests to we handle this in
https://github.com/twilio/twilio-video.js/blob/9a0191e1b311c95454aada7145fa9f1517435a2b/lib/signaling/v2/renderhintssignaling.js#L108
where we forward the request to SFU only if it has changed.

File: lib/media/track/remotevideotrack.js (hunk 1 above)
Comment (makarandp0) [reply lib/media/track/remotevideotrack.js:10]: I think it makes it easy to read when declared locally, but I am open to moving this to constants


Re: 50ms - We decided on 50ms as we are solving only for  re-layout issue - we are not solving for the case where user may scroll up and down quickly. 

The issue we are solving is - when dominant speaker change in our ahoy app (and any apps that follow similar pattern) the position of dominant speaker video changes, it causes a relayout causing the track to be turned off and on quickly. 
our justification for 50ms is here: https://docs.google.com/document/d/11zgCOIYux9vwlNdzd-smrGgnLQf3x0863N-WQ7zzSq8/edit#heading=h.thlab8wykvo

**Review by charliesantos — APPROVED** [review]

**Review by charliesantos — APPROVED** [review]

**Pull request merged**